// rc-hooks/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useSize = exports.usePersistFn = exports.useControllableValue = exports.useClickAway = exports.useUpdateLayoutEffect = exports.useUpdateEffect = exports.useUpdate = exports.useUnmountedRef = exports.useUnmount = exports.useMountedRef = exports.useMount = exports.useSafeState = exports.usePrevious = exports.useSetState = exports.useLimitList = exports.useLatest = exports.useThrottle = exports.useThrottleFn = exports.useDebounce = exports.useDebounceFn = exports.usePagination = exports.useLoadMore = exports.useAsync = void 0;
var tslib_1 = require("tslib");
// Async
var useAsync_1 = require("./useAsync");
Object.defineProperty(exports, "useAsync", { enumerable: true, get: function () { return tslib_1.__importDefault(useAsync_1).default; } });
tslib_1.__exportStar(require("./useAsync"), exports);
var useLoadMore_1 = require("./useAsync/useLoadMore");
Object.defineProperty(exports, "useLoadMore", { enumerable: true, get: function () { return tslib_1.__importDefault(useLoadMore_1).default; } });
tslib_1.__exportStar(require("./useAsync/useLoadMore"), exports);
var usePagination_1 = require("./useAsync/usePagination");
Object.defineProperty(exports, "usePagination", { enumerable: true, get: function () { return tslib_1.__importDefault(usePagination_1).default; } });
tslib_1.__exportStar(require("./useAsync/usePagination"), exports);
// SideEffect
var useDebounceFn_1 = require("./useDebounceFn");
Object.defineProperty(exports, "useDebounceFn", { enumerable: true, get: function () { return tslib_1.__importDefault(useDebounceFn_1).default; } });
var useDebounce_1 = require("./useDebounce");
Object.defineProperty(exports, "useDebounce", { enumerable: true, get: function () { return tslib_1.__importDefault(useDebounce_1).default; } });
var useThrottleFn_1 = require("./useThrottleFn");
Object.defineProperty(exports, "useThrottleFn", { enumerable: true, get: function () { return tslib_1.__importDefault(useThrottleFn_1).default; } });
var useThrottle_1 = require("./useThrottle");
Object.defineProperty(exports, "useThrottle", { enumerable: true, get: function () { return tslib_1.__importDefault(useThrottle_1).default; } });
// State
var useLatest_1 = require("./useLatest");
Object.defineProperty(exports, "useLatest", { enumerable: true, get: function () { return tslib_1.__importDefault(useLatest_1).default; } });
var useLimitList_1 = require("./useLimitList");
Object.defineProperty(exports, "useLimitList", { enumerable: true, get: function () { return tslib_1.__importDefault(useLimitList_1).default; } });
var useSetState_1 = require("./useSetState");
Object.defineProperty(exports, "useSetState", { enumerable: true, get: function () { return tslib_1.__importDefault(useSetState_1).default; } });
var usePrevious_1 = require("./usePrevious");
Object.defineProperty(exports, "usePrevious", { enumerable: true, get: function () { return tslib_1.__importDefault(usePrevious_1).default; } });
var useSafeState_1 = require("./useSafeState");
Object.defineProperty(exports, "useSafeState", { enumerable: true, get: function () { return tslib_1.__importDefault(useSafeState_1).default; } });
// LifeCycle
var useMount_1 = require("./useMount");
Object.defineProperty(exports, "useMount", { enumerable: true, get: function () { return tslib_1.__importDefault(useMount_1).default; } });
var useMountedRef_1 = require("./useMountedRef");
Object.defineProperty(exports, "useMountedRef", { enumerable: true, get: function () { return tslib_1.__importDefault(useMountedRef_1).default; } });
var useUnmount_1 = require("./useUnmount");
Object.defineProperty(exports, "useUnmount", { enumerable: true, get: function () { return tslib_1.__importDefault(useUnmount_1).default; } });
var useUnmountedRef_1 = require("./useUnmountedRef");
Object.defineProperty(exports, "useUnmountedRef", { enumerable: true, get: function () { return tslib_1.__importDefault(useUnmountedRef_1).default; } });
var useUpdate_1 = require("./useUpdate");
Object.defineProperty(exports, "useUpdate", { enumerable: true, get: function () { return tslib_1.__importDefault(useUpdate_1).default; } });
var useUpdateEffect_1 = require("./useUpdateEffect");
Object.defineProperty(exports, "useUpdateEffect", { enumerable: true, get: function () { return tslib_1.__importDefault(useUpdateEffect_1).default; } });
var useUpdateLayoutEffect_1 = require("./useUpdateLayoutEffect");
Object.defineProperty(exports, "useUpdateLayoutEffect", { enumerable: true, get: function () { return tslib_1.__importDefault(useUpdateLayoutEffect_1).default; } });
// Other
var useClickAway_1 = require("./useClickAway");
Object.defineProperty(exports, "useClickAway", { enumerable: true, get: function () { return tslib_1.__importDefault(useClickAway_1).default; } });
var useControllableValue_1 = require("./useControllableValue");
Object.defineProperty(exports, "useControllableValue", { enumerable: true, get: function () { return tslib_1.__importDefault(useControllableValue_1).default; } });
var usePersistFn_1 = require("./usePersistFn");
Object.defineProperty(exports, "usePersistFn", { enumerable: true, get: function () { return tslib_1.__importDefault(usePersistFn_1).default; } });
var useSize_1 = require("./useSize");
Object.defineProperty(exports, "useSize", { enumerable: true, get: function () { return tslib_1.__importDefault(useSize_1).default; } });
