// ut2/es/internals/native.js
var nativeUndefined = void 0;
var stringUndefined = 'undefined';
var stringObject = 'object';
var objectProto = Object.prototype;
var objectProtoToString = objectProto.toString;
var objectProtoHasOwnProperty = objectProto.hasOwnProperty;
var objectProtoPropertyIsEnumerable = objectProto.propertyIsEnumerable;
var objectGetOwnPropertySymbols = Object.getOwnPropertySymbols;
var objectGetPrototypeOf = Object.getPrototypeOf;
var objectKeys = Object.keys;
var functionProto = Function.prototype;
var functionProtoToString = functionProto.toString;
var symbolExisted = typeof Symbol !== stringUndefined;
var symbolProto = symbolExisted ? Symbol.prototype : nativeUndefined;
var arrayProto = Array.prototype;
var arrayProtoSlice = arrayProto.slice;
var mathMin = Math.min;
var mathMax = Math.max;
var mathRandom = Math.random;
var mathFloor = Math.floor;
var mathCeil = Math.ceil;
var mathAbs = Math.abs;
var numberIsFinite = Number.isFinite;
var numberIsInteger = Number.isInteger;
var numberIsSafeInteger = Number.isSafeInteger;
var globalThisExisted = typeof globalThis === stringObject && globalThis;
var globalExisted = typeof global === stringObject && global;
var selfExisted = typeof self === stringObject && self;
var MAX_SAFE_INTEGER = Number.MAX_SAFE_INTEGER || 9007199254740991;
var MIN_SAFE_INTEGER = Number.MIN_SAFE_INTEGER || -9007199254740991;
var MAX_ARRAY_LENGTH = 4294967295;
var bigIntTag = '[object BigInt]';
var numberTag = '[object Number]';
var booleanTag = '[object Boolean]';
var stringTag = '[object String]';
var dateTag = '[object Date]';
var regExpTag = '[object RegExp]';
var symbolTag = '[object Symbol]';
var errorTag = '[object Error]';
var arrayBufferTag = '[object ArrayBuffer]';
var argumentsTag = '[object Arguments]';
var arrayTag = '[object Array]';
var functionTags = ['Function', 'AsyncFunction', 'GeneratorFunction', 'Proxy'].map(function (item) { return '[object ' + item + ']'; });
var weakSetTag = '[object WeakSet]';
var blobTag = '[object Blob]';
var fileTag = '[object File]';
var domExceptionTag = '[object DOMException]';
var objectTag = '[object Object]';
var dataViewTag = '[object DataView]';
var mapTag = '[object Map]';
var promiseTag = '[object Promise]';
var setTag = '[object Set]';
var weakMapTag = '[object WeakMap]';
var windowTag = '[object Window]';

export { MAX_ARRAY_LENGTH, MAX_SAFE_INTEGER, MIN_SAFE_INTEGER, argumentsTag, arrayBufferTag, arrayProto, arrayProtoSlice, arrayTag, bigIntTag, blobTag, booleanTag, dataViewTag, dateTag, domExceptionTag, errorTag, fileTag, functionProto, functionProtoToString, functionTags, globalExisted, globalThisExisted, mapTag, mathAbs, mathCeil, mathFloor, mathMax, mathMin, mathRandom, nativeUndefined, numberIsFinite, numberIsInteger, numberIsSafeInteger, numberTag, objectGetOwnPropertySymbols, objectGetPrototypeOf, objectKeys, objectProto, objectProtoHasOwnProperty, objectProtoPropertyIsEnumerable, objectProtoToString, objectTag, promiseTag, regExpTag, selfExisted, setTag, stringObject, stringTag, stringUndefined, symbolProto, symbolTag, weakMapTag, weakSetTag, windowTag };
