// ut2/lib/internals/helpers.js
'use strict';

var isObjectLike = require('../isObjectLike.js');
var getTag = require('./getTag.js');
var native = require('./native.js');

var VERSION = "1.15.0";
var isBrowser = typeof window !== native.stringUndefined && isObjectLike(window) && typeof document !== native.stringUndefined && isObjectLike(document) && window.document === document;
var supportedArgumentsType = getTag((function () { return arguments; })()) === native.argumentsTag;
var FUNC_ERROR_TEXT = 'Expected a function';
function toSource(func) {
    if (func !== null) {
        try {
            return native.functionProtoToString.call(func);
        }
        catch (e) {
        }
        try {
            return func + '';
        }
        catch (e) {
        }
    }
    return '';
}
var stubFlase = function () { return false; };
var stubTrue = function () { return true; };

exports.FUNC_ERROR_TEXT = FUNC_ERROR_TEXT;
exports.VERSION = VERSION;
exports.isBrowser = isBrowser;
exports.stubFlase = stubFlase;
exports.stubTrue = stubTrue;
exports.supportedArgumentsType = supportedArgumentsType;
exports.toSource = toSource;
