// ut2/lib/internals/native.js
'use strict';

var nativeUndefined = void 0;
var stringUndefined = 'undefined';
var stringObject = 'object';
var objectProto = Object.prototype;
var objectProtoToString = objectProto.toString;
var objectProtoHasOwnProperty = objectProto.hasOwnProperty;
var objectProtoPropertyIsEnumerable = objectProto.propertyIsEnumerable;
var objectGetOwnPropertySymbols = Object.getOwnPropertySymbols;
var objectGetPrototypeOf = Object.getPrototypeOf;
var objectKeys = Object.keys;
var functionProto = Function.prototype;
var functionProtoToString = functionProto.toString;
var symbolExisted = typeof Symbol !== stringUndefined;
var symbolProto = symbolExisted ? Symbol.prototype : nativeUndefined;
var arrayProto = Array.prototype;
var arrayProtoSlice = arrayProto.slice;
var mathMin = Math.min;
var mathMax = Math.max;
var mathRandom = Math.random;
var mathFloor = Math.floor;
var mathCeil = Math.ceil;
var mathAbs = Math.abs;
var numberIsFinite = Number.isFinite;
var numberIsInteger = Number.isInteger;
var numberIsSafeInteger = Number.isSafeInteger;
var globalThisExisted = typeof globalThis === stringObject && globalThis;
var globalExisted = typeof global === stringObject && global;
var selfExisted = typeof self === stringObject && self;
var MAX_SAFE_INTEGER = Number.MAX_SAFE_INTEGER || 9007199254740991;
var MIN_SAFE_INTEGER = Number.MIN_SAFE_INTEGER || -9007199254740991;
var MAX_ARRAY_LENGTH = 4294967295;
var bigIntTag = '[object BigInt]';
var numberTag = '[object Number]';
var booleanTag = '[object Boolean]';
var stringTag = '[object String]';
var dateTag = '[object Date]';
var regExpTag = '[object RegExp]';
var symbolTag = '[object Symbol]';
var errorTag = '[object Error]';
var arrayBufferTag = '[object ArrayBuffer]';
var argumentsTag = '[object Arguments]';
var arrayTag = '[object Array]';
var functionTags = ['Function', 'AsyncFunction', 'GeneratorFunction', 'Proxy'].map(function (item) { return '[object ' + item + ']'; });
var weakSetTag = '[object WeakSet]';
var blobTag = '[object Blob]';
var fileTag = '[object File]';
var domExceptionTag = '[object DOMException]';
var objectTag = '[object Object]';
var dataViewTag = '[object DataView]';
var mapTag = '[object Map]';
var promiseTag = '[object Promise]';
var setTag = '[object Set]';
var weakMapTag = '[object WeakMap]';
var windowTag = '[object Window]';

exports.MAX_ARRAY_LENGTH = MAX_ARRAY_LENGTH;
exports.MAX_SAFE_INTEGER = MAX_SAFE_INTEGER;
exports.MIN_SAFE_INTEGER = MIN_SAFE_INTEGER;
exports.argumentsTag = argumentsTag;
exports.arrayBufferTag = arrayBufferTag;
exports.arrayProto = arrayProto;
exports.arrayProtoSlice = arrayProtoSlice;
exports.arrayTag = arrayTag;
exports.bigIntTag = bigIntTag;
exports.blobTag = blobTag;
exports.booleanTag = booleanTag;
exports.dataViewTag = dataViewTag;
exports.dateTag = dateTag;
exports.domExceptionTag = domExceptionTag;
exports.errorTag = errorTag;
exports.fileTag = fileTag;
exports.functionProto = functionProto;
exports.functionProtoToString = functionProtoToString;
exports.functionTags = functionTags;
exports.globalExisted = globalExisted;
exports.globalThisExisted = globalThisExisted;
exports.mapTag = mapTag;
exports.mathAbs = mathAbs;
exports.mathCeil = mathCeil;
exports.mathFloor = mathFloor;
exports.mathMax = mathMax;
exports.mathMin = mathMin;
exports.mathRandom = mathRandom;
exports.nativeUndefined = nativeUndefined;
exports.numberIsFinite = numberIsFinite;
exports.numberIsInteger = numberIsInteger;
exports.numberIsSafeInteger = numberIsSafeInteger;
exports.numberTag = numberTag;
exports.objectGetOwnPropertySymbols = objectGetOwnPropertySymbols;
exports.objectGetPrototypeOf = objectGetPrototypeOf;
exports.objectKeys = objectKeys;
exports.objectProto = objectProto;
exports.objectProtoHasOwnProperty = objectProtoHasOwnProperty;
exports.objectProtoPropertyIsEnumerable = objectProtoPropertyIsEnumerable;
exports.objectProtoToString = objectProtoToString;
exports.objectTag = objectTag;
exports.promiseTag = promiseTag;
exports.regExpTag = regExpTag;
exports.selfExisted = selfExisted;
exports.setTag = setTag;
exports.stringObject = stringObject;
exports.stringTag = stringTag;
exports.stringUndefined = stringUndefined;
exports.symbolProto = symbolProto;
exports.symbolTag = symbolTag;
exports.weakMapTag = weakMapTag;
exports.weakSetTag = weakSetTag;
exports.windowTag = windowTag;
