// ut2/lib/internals/nodeUtil.js
'use strict';

var native = require('./native.js');

var freeExports = typeof exports === native.stringObject && exports && !exports.nodeType && exports;
var freeModule = freeExports && typeof module === native.stringObject && module && !module.nodeType && module;
var nodeUtil = (function () {
    try {
        var types = freeModule && freeModule.require && freeModule.require('util').types;
        if (types) {
            return types;
        }
    }
    catch (e) {
    }
})();
var nodeIsArrayBuffer = nodeUtil && nodeUtil.isArrayBuffer;
var nodeIsDate = nodeUtil && nodeUtil.isDate;
var nodeIsMap = nodeUtil && nodeUtil.isMap;
var nodeIsRegExp = nodeUtil && nodeUtil.isRegExp;
var nodeIsSet = nodeUtil && nodeUtil.isSet;
var nodeIsTypedArray = nodeUtil && nodeUtil.isTypedArray;

exports.nodeIsArrayBuffer = nodeIsArrayBuffer;
exports.nodeIsDate = nodeIsDate;
exports.nodeIsMap = nodeIsMap;
exports.nodeIsRegExp = nodeIsRegExp;
exports.nodeIsSet = nodeIsSet;
exports.nodeIsTypedArray = nodeIsTypedArray;
